File: src/ros2_cpp_tutorial/src/p2_slow_agent.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
using std::placeholders::_1;

using namespace std::chrono_literals;

/* This example creates a subclass of Node and uses std::bind() to register a
* member function as a callback from the timer. */

class SlowAgent : public rclcpp::Node
{
  public:
    SlowAgent()
    : Node("slow_agent"), count_(0)
    {
      publisher_ = this->create_publisher<std_msgs::msg::String>("slow_topic", 10);
      timer_ = this->create_wall_timer(
        1s, std::bind(&SlowAgent::timer_callback, this));
      subscription_ = this->create_subscription<std_msgs::msg::String>(
        "fast_topic", 10, std::bind(&SlowAgent::topic_callback, this, _1));
    }

  private:
    void timer_callback()
    {
      auto message = std_msgs::msg::String();
      message.data = "Slow timer " + std::to_string(count_++);
      RCLCPP_INFO(this->get_logger(), "[Slow] Publishing: '%s'", message.data.c_str());
      publisher_->publish(message);
    }

    void topic_callback(const std_msgs::msg::String & msg) const
    {
      RCLCPP_INFO(this->get_logger(), "[Slow] I heard: '%s'", msg.data.c_str());
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
    size_t count_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SlowAgent>());
  rclcpp::shutdown();
  return 0;
}

File: src/ros2_python_tutorial/ros2_python_tutorial/p2_slow_agent.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String


class SlowAgent(Node):

    def __init__(self):
        super().__init__('slow_agent')
        self.publisher_ = self.create_publisher(String, 'slow_topic', 10)
        timer_period = 1.0  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.subscription = self.create_subscription(
            String,
            'fast_topic',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning
        self.i = 0

    def timer_callback(self):
        msg = String()
        msg.data = 'Slow timer %d' % self.i
        self.publisher_.publish(msg)
        self.get_logger().info('[Slow] Publishing: "%s"' % msg.data)
        self.i += 1

    def listener_callback(self, msg):
        self.get_logger().info('[Slow] I heard: "%s"' % msg.data)


def main(args=None):
    rclpy.init(args=args)

    slow_agent = SlowAgent()

    rclpy.spin(slow_agent)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    slow_agent.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()